### PR TITLE
Record context edges in the JIT

### PR DESCRIPTION
**Pull request by apaszke**

### CODE REVIEW

File: torch/csrc/jit/init_pass.cpp [hunk 1]
```diff
@@ -46,16 +46,29 @@ void MatchJITOps(std::unique_ptr<Graph>& graph) {
     auto new_op = constructor(graph.get(), p);
     new_op->insertAfter(p);
 
-    if(new_op->hasMultipleOutputs()) {
+    if (new_op->hasMultipleOutputs()) {
+      auto uses = p->uses();
+      for (auto & use : uses) {
+        if (use.user->type()->kind() != TypeKind::HandleType) continue;
```
Comment (zdevito) [inline torch/csrc/jit/init_pass.cpp]: Maybe add a comment that the code that removes Handles is only temporary until we can actually execute with them still in the Graph.

**Review by zdevito — APPROVED** [review]
Looks good.

File: torch/csrc/autograd/functions/special.h [hunk 2]
```diff
@@ -68,8 +68,16 @@ struct Eval : Function {
   static variable_list filterRelevantOutputs(const variable_list& inputs, const variable_list& outputs);
   edge_order computeInputOrder(const variable_list& inputs, const placeholder_list& inherited_placeholders);
 
+  static std::shared_ptr<Eval> getBackwardEval(const variable_list& inputs, const variable_list& outputs) {
+    auto relevant_outputs = filterRelevantOutputs(inputs, outputs);
+    if (relevant_outputs.size() == 0)
+      return nullptr;
+    return std::dynamic_pointer_cast<Eval>(relevant_outputs[0]->grad_fn);
```
Comment (ezyang) [inline torch/csrc/autograd/functions/special.h:75]: EDIT: There used to be a comment here, but then I realized relevant_outputs filters out null outputs.

**Comment by ezyang** [comment]
LGTM. I'd quite like it if there were (1) a test for the "function returns all None, no variables" case and (2) Does filterRelevantOutputs have an invariant that the grad_fns of all variables it returns are the same? If so, we should have an assert to this effect in it. If it doesn't, then the call site of filterRelevantOutputs and then use the 0th element should have an assert for whatever precondition it applied.

File: torch/csrc/jit/init_pass.cpp (hunk 1 above)
Comment (ezyang) [reply torch/csrc/jit/init_pass.cpp]: I don't think this code is temporary. What it says is that if we are replacing a multi-output node with a handle, the replacement is never going to make use of the handle. That seems to be an OK invariant to me. Still worth a comment.

**Comment by ezyang** [comment]
This patch also made me think about https://github.com/ezyang/pytorch/issues/108

**Comment by ezyang** [comment]
Exporter needs to be updated to ignore handles. I'll cook something up.

**Comment by apaszke** [comment]
Ad (2). No, they can be different. Evals can wrap arbitrarily large subgraphs

**Pull request merged**